docs/architecture/adr-110-remote-mempool.md

```diff
@@ -0,0 +1,364 @@
+fact, in this model, full nodes are not intended to interact with the mempool at
+all). DDoS mitigation mechanisms, however, are left to those who implement
+remote mempools.
+
+## Alternative Approaches
+
+The current alternatives considered are:
+
+1. Do nothing, which does not seem sustainable.
+2. Extend ABCI to allow external processes direct access to the P2P layer of the
+   consensus engine, as per [\#1112]. This, however, is a much more complex
+   solution that will require more in-depth discussion and design prior to
+   implementation, as it requires ABCI to allow for _bidirectional_ initiation
+   of requests. This will also require substantial breaking changes to ABCI
+   (primarily for remote applications, since a new transport layer will be
+   required that allows bidirectional initiation of requests, as per [\#1117]).
+3. Provide multiple mempool implementations. This does not seem feasible due to
+   the varying requirements of different applications, including the possibility
+   that different applications need different types of gossip protocols. It is
+   unfeasible capacity-wise for the current CometBFT team to maintain different
+   mempool implementations for specific applications, especially since this
+   falls outside of the purview of maintaining public goods for the _entire_
+   ecosystem as opposed to specific networks.
+
+## Decision
+
+Implement an optional mempool client in Comet that can connect to a remote
+mempool instance/cluster.
+
+## Detailed Design
+
+### High-level architecture
+
+#### Single-process mempool
+
+The simplest possible architecture for a Comet validator with an external
+mempool is as follows.
+
+```mermaid
+flowchart LR
+    comet[Comet Validator]
+    app[Application]
+    mempool[Mempool]
+    user[User]
+
+    comet --> app
+
+    comet -- 4. Reap transactions --> mempool
+    user -- 1. Submit transactions --> app
+    app -- 2. CheckTx --> app
+    app -- 3. Publish transactions --> mempool
+    mempool -- CheckTx --> app
+```
+
+In this flow:
+
+1. Users submit transactions directly to some form of RPC in the application. It
+   is up to operators to secure this RPC endpoint and implement measures to
+   prevent DDoS attacks.
+2. The application validates incoming transactions internally in a similar way
+   to how `CheckTx` currently works.
+3. Transactions that the application deems valid are then sent to the mempool by
+   the application, e.g. through some form of RPC mechanism. The mempool is then
+   expected to propagate these transactions to the rest of the network.
+4. During `PrepareProposal`, a Comet validator will reap transactions from the
+   mempool by way of an RPC call.
+
+Additionally, when transactions are received by the local mempool instance from
+foreign mempool instances, it is expected that the local mempool instance will
+send the transaction to the application for validation (similar to how `CheckTx`
```

Comment:
I am not 100% sure on the terminology here. So let me try to explain what I understand. 

 A remote mempool is remote from CometBFTs points of view, meaning that there is no such thing as a local mempool. The mempool (remote) is propagating transactions to mempools running on other nodes - a `mempool network` if you want it  (which Comet knows nothing about except having RPC endpoints it talks to, but nothing else - txs are not gossiped to the entire p2p network of Comet as far as I understood). When the mempool propagates a transaction to the nodes constituting a mempool network each node receiving a tx will call `CheckTx` .
 
 Is this correct? This sentence made it seem to me as if Comet sends transaction to the local mempool (clist) which then calls `CheckTx` .

Reply:
I agree that this sentence is not very clear. What I understand is that txs are gossiped through the mempool network. And that txs received from (mempool) peers need to be validated by the application. Comet is not involved at all here. And not necessarily a mempool instance is running on the same machine as a Comet instance.

Reply:
"Local" implies the remote mempool instance connected to the local validator. "Foreign" refers to instances connected to foreign validators.